module3/src/MyList.java

```diff
@@ -0,0 +1,697 @@
+     * @throws UnsupportedOperationException if the <tt>set</tt> operation
+     *         is not supported by this list
+     * @throws ClassCastException if the class of the specified element
+     *         prevents it from being added to this list
+     * @throws NullPointerException if the specified element is null and
+     *         this list does not permit null elements
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         (<tt>index &lt; 0 || index &gt;= size()</tt>)
+     */
+    @Override
+    public T set(int index, T element) {
+        if (index < size) {
+            T previous = (T) this.elementData[index];
+            this.elementData[index] = element;
+            return previous;
+        } else {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    /**
+     * Inserts the specified element at the specified position in this list
+     * (optional operation).  Shifts the element currently at that position
+     * (if any) and any subsequent elements to the right (adds one to their
+     * indices).
+     *
+     * @param index index at which the specified element is to be inserted
+     * @param element element to be inserted
+     * @throws UnsupportedOperationException if the <tt>add</tt> operation
+     *         is not supported by this list
+     * @throws ClassCastException if the class of the specified element
+     *         prevents it from being added to this list
+     * @throws NullPointerException if the specified element is null and
+     *         this list does not permit null elements
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         (<tt>index &lt; 0 || index &gt; size()</tt>)
+     */
+    @Override
+    public void add(int index, Object element) {
+        throw new UnsupportedOperationException();//return;  //Stanislav said we do not need to implement it
+    }
+
+    /**
+     * Removes the element at the specified position in this list (optional
+     * operation).  Shifts any subsequent elements to the left (subtracts one
+     * from their indices).  Returns the element that was removed from the
+     * list.
+     *
+     * @param index the index of the element to be removed
+     * @return the element previously at the specified position
+     * @throws UnsupportedOperationException if the <tt>remove</tt> operation
+     *         is not supported by this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         (<tt>index &lt; 0 || index &gt;= size()</tt>)
+     */
+    @Override
+    public T remove(int index) {
+        if (index < 0 || index >= size || size < 1) {
```

Comment:
size < 1 is not possible. Understand - why?

Reply:
We have 2 methods which can decrement `size` – `remove(int index)` and `remove()`.
`remove(int index)`  will decrement `size` only if `size` >= 1, and will throw exception in case when `size`=0 even without this qualification. 
`remove()` will decrement `size` only if `size` >= 1 (it is influenced by `currentIndex` which can be >= 0 only)

That is how I understand.

Reply:
take a look only on the logic of the if `index < 0 || index >= size || size < 1`
1. index < 0 - other verification will be skipped and exception thrown
2. index >= size - here we understand that index > 0, because of previous verification, so then it makes that - size >= 1
3. size < 1 - it's impossible because above verification. Since size >= 1 and cannot be les then 1